### PR TITLE
VM Support Refinement

### PR DESCRIPTION
**Pull request by jrajahalme**
Split IPAM bootstrap into configuration and startup phases so that the configuration can still be modified in between, and join cluster (if --join-cluster is set) before starting IPAM. This allows cluster to provide cluster CIDR config for the VM.

Add CiliumExternalWorkload CRD that dictates the namespace and labels for an external workload, such as a VM. VM's hostname and CEW CRD name must match, and the VM is not admitted if a matching CEW CRD is not found.

Move 'hubble.tls.auto.cronJob' values to 'certgen' and add a 'hubble.' prefix to hubble specific helm partials.

Add helm templates and values for `clustermesh-apiserver`.

Move `cilium/clustermesh-apiserver` to the main Cilium repo.

Add `PROVISION_EXTERNAL_WORKLOAD` option to compile and run Cilium agent in docker in the runtime CI VM. This mirrors the planned deployment model for Cilium Agent in external workload VMs.

TODO:
- [x] Update `cilium/certgen` version and SHA after required changes have been merged. Before this `.clustermesh.apiserver.tls.auto.method='cronJob'` will fail.
- [x] Enable external workloads in `install/kubernetes/experimental-install.yaml` (depends on the `certgen` update)
- [ ] CI test(s)

For later PRs:
- [x] validate k8s service access from VMs
- [ ] test with direct routing modes, remove the need for IPAM setup in this case
- [ ] check if tunneling can be done without any IPAM (i.e., w/o creating `cilium_host` device)
- [x] wire cluster DNS to the VMs so that VMs can use cluster-defined DNS names
- [ ] pass policy from cluster to the VMs
- [ ] enforce policy in VMs

Fixes: #13309

### CODE REVIEW

**Comment by jrajahalme** [comment]
Test setup consists of two VMs, one running with k8s (`k8s1`), one running without k8s (`runtime`). Setup steps:

1. Clone `https://github.com/isovalent/clustermesh-apiserver.git`at parent directory of your `cilium` repo. After this the parent should have both `cilium` and `clustermesh-apiserver` directories.
2. Check out `pr/jrajahalme/vm-scoping` from `cilium/cilium`
3. `cd cilium/test`
4. Start `k8s1`:
`SHARE_PARENT=1 K8S_NODES=1 ./vagrant-local-start.sh`
 - This may take a long time for the 1st run, as a vagrant box is packaged (into `test/.vagrant/k8s-box-package.box`). Subsequent runs will be faster. This packaging removes the need to pull test images for later runs of `./vagrant-local-start.sh`.
5. `vagrant ssh k8s1-1.19`
6. `cd ~/go/src/github.com/cilium/cilium`
7. `DOCKER_BUILDKIT=1 DOCKER_IMAGE_TAG=testing make docker-operator-generic-image dev-docker-image`
8. Apply the following patch to `install/kubernetes/quick-install.yaml`:
```
diff --git a/install/kubernetes/quick-install.yaml b/install/kubernetes/quick-install.yaml
index a1819136ce..99f988ebde 100644
--- a/install/kubernetes/quick-install.yaml
+++ b/install/kubernetes/quick-install.yaml
@@ -353,6 +353,7 @@ spec:
             topologyKey: kubernetes.io/hostname
       containers:
       - args:
+        - --debug
         - --config-dir=/tmp/cilium/config-map
         command:
         - cilium-agent
@@ -424,8 +425,8 @@ spec:
               key: custom-cni-conf
               name: cilium-config
               optional: true
-        image: "docker.io/cilium/cilium:latest"
-        imagePullPolicy: Always
+        image: "docker.io/cilium/cilium-dev:testing"
+        imagePullPolicy: IfNotPresent
         lifecycle:
           postStart:
             exec:
@@ -622,8 +623,8 @@ spec:
               key: debug
               name: cilium-config
               optional: true
-        image: "docker.io/cilium/operator-generic:latest"
-        imagePullPolicy: Always
+        image: "docker.io/cilium/operator-generic:testing"
+        imagePullPolicy: IfNotPresent
         name: cilium-operator
         livenessProbe:
           httpGet:
```
9. `k apply -f install/kubernetes/quick-install.yaml`
10. `cd ~/go/src/github.com/cilium/clustermesh-apiserver`
11. `git fetch --all && git checkout origin/pr/jrajahalme/gke-fixes`
12. `export DOCKER_BUILDKIT=1; docker build -t cilium/clustermesh-apiserver:testing .`
13. `ks apply -f install/`
14. Observe the IP address of `clustermesh-apiserver` POD:
`kubectl get pods -n kube-system -o jsonpath='{.items[?(@.metadata.labels.name=="clustermesh-apiserver")].status.podIP}{"\n"}'`
15. Change to `test` directory and start the `runtime` VM:
`PROVISION_EXTERNAL_WORKLOAD=1 SHARE_PARENT=1 ./vagrant-local-start-runtime.sh`
16. `vagrant ssh runtime`
17. Check that etcd has connected successfully, observe the first line of:
`cilium status`
You should see something like:
`etcd: 1/1 connected, lease-ID=7c02748328e75f57, lock lease-ID=7c02748328e75f59, has-quorum=true: 192.168.36.11:30367 - 3.4.7 (Leader)`
18. Ping the IP of `clustermesh-apiserver` using the POD IP from step 11 above:
`ping 10.0.0.X`
 19. The ping should keep running also when the following CCNP is applied in `k8s1` (e.g., `kubectl apply -n kube-system -f test-ccnp.yaml`):
```
apiVersion: cilium.io/v2
description: TEST CCNP
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: test-ccnp
  namespace: kube-system
spec:
  endpointSelector:
    matchLabels:
      name: clustermesh-apiserver
  ingress:
  - fromEndpoints:
    - matchLabels:
        io.kubernetes.pod.name: runtime
  - toPorts:
    - ports:
      - port: "2379"
        protocol: TCP
```

File: daemon/cmd/ipam.go
Comment (tklauser) [inline daemon/cmd/ipam.go]: I think this should be removed (as well as the corresponding `bootstrapStats.ipam.End(true)` at line 363) since the functionality is moved to `(*Daemon).startIPAM()` by this PR.

File: pkg/node/store/store.go
Comment (tklauser) [inline pkg/node/store/store.go]: Nit:

```suggestion
	if err != nil {
		registerStore.Release()
		return nil, err
	}

	// Wait until an updated key is received from the kvstore
	select {
	case n = <-registerObserver.updates:
	case <-time.After(defaults.NodeInitTimeout / 10):
		registerStore.Release()
		return nil, fmt.Errorf("timed out waiting for node identity")
	}
```

File: daemon/cmd/daemon.go
Comment (aanm) [inline daemon/cmd/daemon.go]: log.WithFields.. for the `nodeTypes.GetName()`

File: pkg/node/types/node.go
Comment (aanm) [inline pkg/node/types/node.go]: Can be squashed with the previous commit.

**Review by aanm — APPROVED** [review]
Nothing blocking from my side

**Comment by aanm** [comment]
@jrajahalme which entities will access the new CRD? Shouldn't we add it to the Cilium's cluster role?

File: examples/crds/ciliumexternalworkloads.yaml
Comment (christarazi) [inline examples/crds/ciliumexternalworkloads.yaml:1]: Nit: you can squash this commit with the previous commit that makes changes to the CRD. It makes sense to combine, in case someone `git blame`s on the generated file, they'll be lead to the "wrong" commit.

File: pkg/k8s/apis/cilium.io/v2/cew_types.go
Comment (christarazi) [inline pkg/k8s/apis/cilium.io/v2/cew_types.go:1]: You'll need to add two more things:

1) Make the agent wait for this new CRD, see commit e91f803c3
2) As Andre mentioned, you'll need to add the RBAC permissions for the agent to access this new resource, see https://github.com/cilium/cilium/blob/17c4f80ea32495ecd62c4f1a39455c45b2210f6c/install/kubernetes/cilium/templates/cilium-agent-clusterrole.yaml#L90

**Review by christarazi — CHANGES_REQUESTED** [review]
Mostly reviewed K8s side of things; looks good besides a few items to address.

File: pkg/k8s/apis/cilium.io/v2/cew_types.go
Comment (jrajahalme) [reply pkg/k8s/apis/cilium.io/v2/cew_types.go:1]: Cilium agents are not accessing CiliumExternalWorkload CRDs, so IMO agent RBAC permissions do not need to be amended. Only `clustermesh-apiserver` reads CEW CRDs :-)

File: daemon/cmd/ipam.go
Comment (jrajahalme) [reply daemon/cmd/ipam.go]: Done.

File: pkg/node/store/store.go
Comment (jrajahalme) [reply pkg/node/store/store.go]: Applied, thanks!

File: daemon/cmd/daemon.go
Comment (jrajahalme) [reply daemon/cmd/daemon.go]: Changing this to debug as well.

File: pkg/node/types/node.go
Comment (jrajahalme) [reply pkg/node/types/node.go]: I lost the context for this. But in case this is for the first two commits (where `pkg/node/types/node.go`) is changed, I'll keep them separate until we get clean CI runs so that triaging will be easier if the IPAM bootstrap changes cause problems.

File: examples/crds/ciliumexternalworkloads.yaml
Comment (jrajahalme) [reply examples/crds/ciliumexternalworkloads.yaml:1]: Will do. I kept them separate so that I had a clean view which files I'm working with and which are merely generated :-)

File: pkg/k8s/apis/cilium.io/v2/cew_types.go
Comment (jrajahalme) [reply pkg/k8s/apis/cilium.io/v2/cew_types.go:1]: @christarazi Also, I'm not sure if `clustermesh-apiserver` needs to wait for CRDs or not? It is watching on a number of CRDs, as well as creating and updating CRDs.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Test drove the testing instructions, `clustermesh-apiserver/test/testccnp.yaml` needs a fix, correct contents is shown above. (will wait for CI run to finish before pushing the fixed policy file.)

File: pkg/node/types/node.go
Comment (jrajahalme) [reply pkg/node/types/node.go]: Squashed the first two commits together.

File: pkg/k8s/apis/cilium.io/v2/cew_types.go
Comment (aanm) [reply pkg/k8s/apis/cilium.io/v2/cew_types.go:1]: @jrajahalme If that's the case, cilium agent does not need any RBAC permissions to access the new resource. However, `clustermesh-apiserver` needs those RBAC permissions.

>@christarazi Also, I'm not sure if `clustermesh-apiserver` needs to wait for CRDs or not? It is watching on a number of CRDs, as well as creating and updating CRDs.

It depends, if `clustermesh-apiserver` depends on those CRDs then it needs to be sure they are available, and if not, it needs to wait for them.

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/k8s/apis/cilium.io/v2/cew_types.go
Comment (jrajahalme) [reply pkg/k8s/apis/cilium.io/v2/cew_types.go:1]: OK, I'll need to add a wait for the CRDs before starting any of the CRD watchers in `clustermesh-apiserver`.

**Review by gandro — APPROVED** [review]
I only reviewed the changes regarding `certgen` and `hubble.tls.auto` helm changes. They look good to me!

**Review by aanm — APPROVED** [review]

File: install/kubernetes/cilium/values.yaml
Comment (christarazi) [inline install/kubernetes/cilium/values.yaml]: This is probably not supposed to be included?

**Review by christarazi — APPROVED** [review]
LGTM assuming no further changes to this PR (i.e. CRD watching for the clustermesh-apiserver). Just one minor comment to resolve, doesn't require a second review.

File: install/kubernetes/cilium/values.yaml
Comment (jrajahalme) [reply install/kubernetes/cilium/values.yaml]: Right, will remove!

Comment (jrajahalme) [reply install/kubernetes/cilium/values.yaml]: Fixed!

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Added waiting for CRDs to `clustermesh-apiserver`.

File: pkg/k8s/synced/crd.go
Comment (christarazi) [inline pkg/k8s/synced/crd.go]: I think we can rename these now to simply `add` and just drop the `CRD` part completely as it's redundant.

File: pkg/k8s/synced/crd.go
Comment (christarazi) [inline pkg/k8s/synced/crd.go:260]: I think this is fine for explicit-ness, but not necessary. Just wondering, did you run into trouble with this or is this change just for "readability"?

File: pkg/k8s/synced/resources.go
Comment (christarazi) [inline pkg/k8s/synced/resources.go:27]: Nit: godoc

File: pkg/k8s/synced/crd.go
Comment (christarazi) [inline pkg/k8s/synced/crd.go]: Why is CCLRP also being added?

Comment (christarazi) [reply pkg/k8s/synced/crd.go]: I believe it is also being removed in a pending PR: https://github.com/cilium/cilium/pull/13671/commits/33e8399a806fe217606721ca87118819ef16a20b

File: clustermesh-apiserver/main.go
Comment (christarazi) [inline clustermesh-apiserver/main.go]: Is the `CRDWaitTimeout` supposed to be used to create a context and passed in here?

File: pkg/k8s/synced/crd.go
Comment (christarazi) [inline pkg/k8s/synced/crd.go:17]: I think the refactor is good. I think a package name like `resources` would be more apt to represent the code inside it. What do you think?

**Review by christarazi — CHANGES_REQUESTED** [review]
Overall looks good. Some items to iterate on.

File: pkg/k8s/synced/crd.go
Comment (christarazi) [inline pkg/k8s/synced/crd.go]: Maybe we can consider renaming this function to `SyncCRDs` and having the following godoc:

```
SyncCRDs will sync Cilium CRDs to ensure that they have all been installed inside the K8s cluster. These CRDs are added by the Cilium Operator. This function will block until it finds all the CRDs or if a timeout occurs.

File: pkg/k8s/synced/crd.go
Comment (christarazi) [inline pkg/k8s/synced/crd.go]: Nit: Let's move this up in the file since it's exported. Generally, exported functions should come first, and then unexported / private APIs can follow.

File: pkg/k8s/synced/resources.go
Comment (christarazi) [inline pkg/k8s/synced/resources.go]:
```
// WaitForCacheSync waits for all K8s resources represented by resourceNames to have their K8s caches synchronized. 
```

**Review by christarazi — CHANGES_REQUESTED** [review]

File: pkg/k8s/synced/crd.go
Comment (jrajahalme) [reply pkg/k8s/synced/crd.go]: Done, changed "deleteCRD" to "remove" as well.

File: pkg/k8s/synced/crd.go
Comment (jrajahalme) [reply pkg/k8s/synced/crd.go:260]: Right, was not absolutely necessary, but thought keeping three places where the CRDs were listed "in sync" is unnecessary when can have it done only at one location.

Comment (jrajahalme) [reply pkg/k8s/synced/crd.go:260]: So more for "maintainability" :-)

File: pkg/k8s/synced/resources.go
Comment (jrajahalme) [reply pkg/k8s/synced/resources.go:27]: Added.

File: clustermesh-apiserver/main.go
Comment (christarazi) [reply clustermesh-apiserver/main.go]: I'm also not entirely clear why we need to also declare the `CRDWaitTimeout` flag again in this file.

File: pkg/k8s/synced/crd.go
Comment (jrajahalme) [reply pkg/k8s/synced/crd.go]: Only saw it was missing, was not aware of the PR for removing it. Will remove :-)

File: clustermesh-apiserver/main.go
Comment (jrajahalme) [reply clustermesh-apiserver/main.go]: Seems not needed, as there is a fatal log upon CRDWaitTimeout within the waiter code. Bumped into it before I added the code to use the `CRDWaitTimeout` option.

File: pkg/k8s/synced/crd.go
Comment (jrajahalme) [reply pkg/k8s/synced/crd.go:17]: My reading of "resources" has a connotation that it would include definition or more of the resource management... This is just a simple tracker for the initial resource sync. Started with `sync` as the package name, but soon figured that was pretty bad :-)

Comment (christarazi) [reply pkg/k8s/synced/crd.go:17]: Fair enough.

> This is just a simple tracker for the initial resource sync

I think this is a good start to a package-level godoc. Could you create a `doc.go` and include a godoc including the above quoted text? I agree, naming this is kinda hard, so having a godoc to explain it will be valuable. Something like:

```
Package synced provides a simple tracker to ensure that Cilium K8s resources such as CRDs are synchronized in the cluster. It also ... blah blah (if there's a need to expand more...)
```

File: pkg/k8s/synced/crd.go
Comment (jrajahalme) [reply pkg/k8s/synced/crd.go]: Done, thanks!

File: pkg/k8s/synced/crd.go
Comment (jrajahalme) [reply pkg/k8s/synced/crd.go]: Done.

File: pkg/k8s/synced/resources.go
Comment (jrajahalme) [reply pkg/k8s/synced/resources.go]: Changed.

File: pkg/k8s/synced/crd.go
Comment (jrajahalme) [reply pkg/k8s/synced/crd.go:17]: I'll leave this as-is for now, but don't oppose potential future changes for a better name.

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/k8s/synced/logfields.go
Comment (christarazi) [inline pkg/k8s/synced/logfields.go]: Nit: 2020

**Review by christarazi — APPROVED** [review]
LGTM! Just a few more non-blocking items and they don't require rerunning CI, so let's wait until CI finishes.

- ~CRDWaitTimeout is still being declared in clustermesh main.go~
- Package level godoc on package `synced`

**Comment by jrajahalme** [comment]
retest-gke

File: pkg/k8s/synced/logfields.go
Comment (jrajahalme) [reply pkg/k8s/synced/logfields.go]: Docs fixed before the merge

**Pull request merged**
